Python clean up for both python 2 and 3

danlangford opened the pull request:
when trying to run the python i found some scripts only ran with 2 and others only ran with 3. i found some simple improvements to make, some compatibility issues to address, some style clean up to perform. the code now more generally conforms to PEP8. i think there are a couple issues this could address

edit: see the latest take on this PR here https://github.com/buttonmen-dev/buttonmen/pull/2667#issuecomment-972614387

danlangford commented:
This primarily fixes #2197

danlangford commented:
it should be noted that this is (should be) python 2 and 3 compatible but I really question if we need to maintain python 2 compatibility. 

There are some compatibility layers. Initially I stayed on "urllib" and used the package "future" with "install_aliases" for python 2 compatibility but ran into a bug I could not fix. Most searching the community suggests busting using the higher level "requests" library. POINT IS there is now a requirements.txt file 

Now I know this was handled initially with the try-catch imports. I had a goal to remove those and make the code feel like python 3 with backwards compatibility and not be littered with little hacks. This is generally the direction of the python community. https://python-future.org/compatible_idioms.html

python support will cease in many major products this month.

irilyth commented:
I don't feel strongly about it, but my two cents is that we could drop Python 2 support.

danlangford commented:
what are some things i can do to validate that some of this is still functional? i see some replay loops and random_ai and i wonder if somebody can give me some direction. i feel fairly confident that `bmapi.py` and `bmutils.py` are good.

cgolubi1 commented:
Apologies for being so slow.  I will get back to you as soon as i can --- really hoping to have some time next week.  The replay stuff is the regression test code i use to, basically, play a large number of random games and look for logic bugs and unexpected API regressions.  It needs to keep working, but it's fine for us to update it so it does, since the only customer for it is replay-testing itself.  I'll take a look ASAP!

danlangford commented:
i dont mean to push anybody too fast. take your time. thanks. im still pouring through it and may make some more commits to officially drop python2 support which will allow my linter to be a bit more aggressive on its suggestions for improvements

cgolubi1 commented:
Okay, i've looked at this at all for the first time.  I feel mildly negative about reformatting random_ai.py.  It makes my life a little harder because i feel obligated to maintain a file in a style i didn't pick and am not going to naturally generate, and it helps nobody, because nobody other than me uses that file.

Does anyone have a non-philosophical benefit to reformatting that file?  As in, something that doing so will help you personally accomplish?  If not, i'd say let's skip that one.

cgolubi1 commented:
In general: i'd appreciate a refresher in what your goal is in modifying this code.  Right now we have a bit of a jumble of things in the python directory, and i'd say that fixing up things like bmapi and bmutils, and the various helper scripts, so that anyone can use them, makes a lot of sense, whereas making modifications to the replay game stuff which is only for regression testing, doesn't make much sense.

I think it'd be fine for us to clean up the directory structure as a way to differentiate between stuff that is designed for general player use and development, and stuff that is really just for the replay testing use case.  I'm very happy to look at PRs that do that, or respond to comments suggesting how you might want to do that.

danlangford commented:
This sort of feedback is exactly what I am looking for @cgolubi1 thank you for the tone you took to consider this ideas. 

1) can somebody help me identify what scripts are for regression testing and what are more for people to generally dev with and any other "category" of python scripts

2) in regards to code style. One benefit of having a structured code style is that many modern IDEs will be be able to help make sure your code can conform to the style of your choice and then all your files are consistent. It can help when people move between files to have a common style. Many projects that have lots of contributors will enforce a code style like PEP8 so that it doesn't get too messy. I prefer PEP8 with 2 space indents instead of 4. However core contributor productivity should take a higher priority than a code style you do not agree with. PEP8 is very common or the Google pylint which based on PEP8 but a little tighter 

3) with your permission I will gladly make the bmapi and bmutils a bit more user friendly for other devs to play with 

4) @cgolubi1 are you ok with dropping python 2 and relying on python 3?

cgolubi1 commented:
Here's the breakdown of the scripts and libraries that exist right now: i think they all fall into one of the categories of USER or REPLAYTEST

tools/api-client/python/:
buttons-with-skills* - USER
create_game* - USER
game-create* - USER
game-data* - USER
game_info* - USER
head-to-head* - USER
interactive_mode.py - USER
monitor* - USER
prep_replay_games* - REPLAYTEST
record* - USER
replay_loop* - REPLAYTEST
replay_single_game* - REPLAYTEST
test_log_games* - REPLAYTEST
update_replay_games* - REPLAYTEST

tools/api-client/python/lib/:
bmapi.py - USER (used by REPLAYTEST)
bmutils.py
random_ai.py*

danlangford commented:
to clarify, my primary goal is to bring the python up to a modern expectation because I have been doing lots of python development around buttonmen and i have already done the work and would like to contribute to it. 

some of the python work i have been doing includes some MMR Leaderboards for one of my tournaments on the site as well as writing an adapter for BMAI.

cgolubi1 commented:
As for 2 --- totally understood, and i'm sure i'll get there someday, but doesn't solve a problem i have right now wrt the replay scripts.  I think we're on the same page.

As for 3 --- sounds great, go for it.

As for 4 --- philosophically, i have no problem with this.  In practice, when you commit a bmapi.py that doesn't support python 2, i'll have to update my replay rig to run under python3.  (As noted by the above division, i think bmapi.py is the only script this applies to.)

blackshadowshade commented:
Considering that this hasn't advanced for a while, I'd suggest that @cgolubi1 clones this branch and adjusts it to meet her requirements, and then submits a new pull request, to be reviewed by @danlangford. I think that would help us to meet both sets of expectations in the fewest iterations.

danlangford commented:
if @cgolubi1 is open to that it's fine. However @cgolubi1 wasn't the one who wanted a change here so it's not really their responsibility. 

@blackshadowshade you likely want to close the ticket? My feelings won't be hurt if the ticket is closed as "won't do"

danlangford commented:
> I think it'd be fine for us to clean up the directory structure as a way to differentiate between stuff that is designed for general player use and development, and stuff that is really just for the replay testing use case. I'm very happy to look at PRs that do that, or respond to comments suggesting how you might want to do that.

@cgolubi1 the current dir structure inside `./tools/` is this:

```
.
├── README
└── api-client
    ├── README
    └── python
        ├── lib
        │   ├── bmapi.py
        │   ├── bmutils.py
        │   └── random_ai.py
        ├── README.md
        ├── bmrc.example
        ├── buttons-with-skills
        ├── create_game
        ├── game-create
        ├── game-data
        ├── game_info
        ├── head-to-head
        ├── interactive_mode.py
        ├── monitor
        ├── prep_replay_games
        ├── record
        ├── replay_loop
        ├── replay_single_game
        ├── test_log_games
        └── update_replay_games
```

so we have a dir that is called `api-client` but in addition to some api client classes it also contains replay tooling and user scripts. So I agree this could use some cleaning up. 

a couple of ideas:

```
.
├── README
└── api-client
    ├── README
    └── python
        ├── README.md
        ├── bmrc.example
        ├── bots
        │   └── random_ai.py
        ├── lib
        │   ├── bmapi.py
        │   └── bmutils.py
        ├── replaytest
        │   ├── prep_replay_games
        │   ├── replay_loop
        │   ├── replay_single_game
        │   ├── test_log_games
        │   └── update_replay_games
        └── scripts
            ├── buttons-with-skills
            ├── create_game
            ├── game-create
            ├── game-data
            ├── game_info
            ├── head-to-head
            ├── interactive_mode.py
            ├── monitor
            └── record
```

or 

```
.
├── README
└── python
    ├── api-client
    │   ├── bmapi.py
    │   └── bmutils.py
    ├── random_ai
    │   └── random_ai.py
    ├── replaytest
    │   ├── prep_replay_games
    │   ├── replay_loop
    │   ├── replay_single_game
    │   ├── test_log_games
    │   └── update_replay_games
    └── user
        ├── buttons-with-skills
        ├── create_game
        ├── game-create
        ├── game-data
        ├── game_info
        ├── head-to-head
        ├── interactive_mode.py
        ├── monitor
        └── record
```


thoughts?

blackshadowshade commented:
> if @cgolubi1 is open to that it's fine. However @cgolubi1 wasn't the one who wanted a change here so it's not really their responsibility.
> 
> @blackshadowshade you likely want to close the ticket? My feelings won't be hurt if the ticket is closed as "won't do"

My feel is that @cgolubi1 would be open to accepting most of this, but may simply want to exclude a little bit of the content. However, I have no real idea how much work has already been done and how much remains to be done.

I also don't mind how you two iterate on this pull request, it doesn't affect me directly.

irilyth commented:
In the unlikely event that anyone cares what I think, I continue to think that it'd be fine to drop support for Python 2, at least for the general player use stuff.

danlangford commented:
ok i have been working on this a bit and would like to provide a summary of this ENTIRE PR, top to bottom

# GOALS
- allow users to execute scripts with Python 3
- allow the `replaytests` to remain on their current Python 2 tooling
- where reasonable trend the python towards community standards

-------

# CODE CHANGES
- use of `__future__` and `future` imports for `bmapi` to support Py2 and Py3 with heavy influence from https://python-future.org/compatible_idioms.html
- some files have been reformatted to be more PEP8 style compliant. 
- replaces use of some modules marked for deprecation (i.e. `argparse`)
- some Python 2 had gentle adjustments to encourage small trends towards Python 3 (i.e. `print()` and `raise ValueError()`)
- If my IDE pointed out blatant errors they have also been fixed (i.e. `if player == None` -> `if player is None`)

# LIBRARY/MODULE USE
- added external libraries/modules only where deemed extremely valuable
  - > `urllib` is the hardest module to use from Python 2/3 compatible code. You might want to switch to Requests instead. 
- the `requests` library is likely the most notable case of this in `bmapi.py`
- it should be noted that pip is being used to download those third party dependencies. `requirements.txt` files exist to help you quickly get those (i.e. `pip install -r requirements.txt`)

# STRUCTURE
- restructured the code to more accurately organize all the various pythons scripts and their target audience/usage. Heavy influence from https://realpython.com/python-application-layouts/#installable-single-package
- added `__init__.py` files in some places so Python could do a cleaner job of importing those directories as "modules"

```
.
├── README
└── python
    ├── api-client
    │   ├── __init__.py
    │   ├── bmapi.py
    │   └── bmutils.py
    ├── random_ai
    │   ├── __init__.py
    │   └── random_ai.py
    ├── replaytest
    │   └── # replaytest scripts here
    └── user
        └── # user scripts here
```

# DOCUMENTATION
- added or enhanced various README files around the directory structure to inform intent and guide usage
- add many PEP8 style python doc strings

# TESTING
- added Puppet module `buttonmen::clienttestsetup` to ensure Py2 and Py3 are installed along with any python libraries needed to run
- CircleCI steps added to run `test.bmapi.py` with Py2 and Py3
- NOTE: I have been unable to fully test that the vagrant/virtualbox tooling runs these tests just fine
- CircleCI tests are passing


im interested in thoughts from anybody especially @cgolubi1  and @irilyth

cgolubi1 commented:
Thanks for the detailed description, Dan.  As soon as i can, i'll at least setup a replay loop so i can get some data about whether the scripts work for replay-testing, and take it from there.

danlangford commented:
@cgolubi1 be aware that I dont think it will work 100% out of the box. bmapi now depends on a 3rd party library that you will need to have installed. my hope is that your replaytest set up has `pip` installed and that you could incorporate calling `pip install -r ../path/to/api-client/requirements.txt` obv with the correct path based on where you run the replaytest suite from

danlangford commented on deploy/vagrant/modules/buttonmen/manifests/init.pp:
```diff
@@ -129,3 +129,28 @@
   }
 }
+
+class buttonmen::clienttestsetup {
+
+  package {
+    "python": ensure => installed;
+    "python3": ensure => installed;
+  }
+
+  exec {
+
+    "pip_install":
+      command     => "/usr/bin/wget -qO - https://bootstrap.pypa.io/pip/2.7/get-pip.py | /usr/bin/python",
+      require => [Package["wget"], Package["python"]];
+    "pip3_install":
+      command     => "/usr/bin/wget -qO - https://bootstrap.pypa.io/pip/3.5/get-pip.py | /usr/bin/python3",
+      require => [Package["wget"], Package["python"]];
+
+    "pip_requirements_install":
+      command     => "/usr/bin/python -m pip install -r /buttonmen/tools/python/api_client/requirements.txt",
+      require => [Exec["pip_install"]];
+    "pip3_requirements_install":
+      command     => "/usr/bin/python3 -m pip install -r /buttonmen/tools/python/api_client/requirements.txt",
+      require => [Exec["pip3_install"]];
+  }
+}
```
maybe I shouldn't have done this client test setup in puppet as it then installs these tools in more places then it needs to be. maybe I should have done these as a step in circleci since that's the only place the actually needs these tests to run? ¯\_(ツ)_/¯ im looking for others who have more experience with designing the current vagrant/puppet/circleci/docker set up to weigh in and confirm or correct my decisions here

cgolubi1 replied:
I agree with the decision to do this in puppet, if we're going to add more mandatory dependencies.  It *might* mean this stuff works out of box for dev and replay sites, certainly makes it more likely.

irilyth commented:
> im interested in thoughts from anybody especially @cgolubi1 and @irilyth

I haven't done anything with any of this stuff in a while, and am relatively content to see it change; if that ends up requiring me to update my stuff that uses it, that should be fine.

cgolubi1 commented:
Just to state where i am on this: i think python2/python3 compatibility is actively good.  Rearranging all the files is a short-term negative for me, because it means i have to put in some work to update the replay test harness to work with the new ordering.  But i'm willing to accept it as the price i have to pay to get the python2/python3 work done, and i don't actually care where the files sit --- i just *use* where the files sit when building replay sites, so when it changes, that's an interface i have to update.

I'm working on that now as time permits, and i'll keep y'all posted --- this comment is basically just a caution/request to make sure the new directory layout is one you like, because it's not *that* trivial to change it, so once it's changed, i'll want to stick with the new layout for awhile.

cgolubi1 commented:
You said:

> the requests library is likely the most notable case of this in bmapi.py
>    this did impact the cookiejar creation and management

That's indeed the case!  After my initial path munging, i'm now stuck at:

```python
$ ./replay_loop 
mysqlshow: Unknown database 'buttonmen'
Creating buttonmen
Populating buttonmen
buttonmen_test already exists - recreating it
Creating buttonmen_test
Populating buttonmen_test
mysql: [Warning] Using a password on the command line interface can be insecure.
Traceback (most recent call last):
  File "./test_log_games", line 34, in <module>
    if not b1.login():
  File "../api_client/bmapi.py", line 98, in login
    self.cookiejar.save(ignore_discard=True)
AttributeError: BMClient instance has no attribute 'cookiejar'
```

The next step is for Dan to provide a config syntax example of how i should work with the new cookie jar.  Once i have that, i'll check it out, and keep working on getting this to run.

cgolubi1 commented:
FWIW, right now i have:

```
$ cat ~/share/buttonmen/.bmrc 
[localr3]
url=http://localhost/api/responder
username=responder003
password=<password>
cookiefile=/home/chaos/share/buttonmen/.bmcookies.local.r3

[localr4]
url=http://localhost/api/responder
username=responder004
password=<password>
cookiefile=/home/chaos/share/buttonmen/.bmcookies.local.r4

```

That seems like it should still be compatible with the example bmrc in this PR:

```
$ cat ./tools/python/bmrc.example
##### Example .bmrc file
# Move this to .bmrc and fill in username/password for any sites
# you want to use

[dev]
url=http://dev.buttonweavers.com/api/responder
username=
password=
cookiefile=.bmcookies.dev

[www]
url=http://www.buttonweavers.com/api/responder
username=
password=
cookiefile=.bmcookies.www

```

danlangford commented:
sorry I wasn't more responsive on this. are things still stuck? the config should remain the same but it looks like the replay_loop scripts interactions with BMClient.cookiejar needs to be enhanced. I tried to search through the replay scripts but didn't feel confident enough in them to execute them. how can I help?

cgolubi1 commented:
Hmm, okay.  Is there a good example somewhere else in the codebase of how python code that uses BMClient *should* look?  `test_log_games` doesn't interact with cookiejar directly, nor should it afaik.  It pretty much just runs:

```
b1 = bmapi.BMClient(CONFIGPATH, 'localr3')
b2 = bmapi.BMClient(CONFIGPATH, 'localr4')
b1.login()
...
```

and that `b1.login()` is what's blowing up.

Can you point me to one of the scripts that you're confident works out of the box with python2, and i'll see what's different between that and what i'm doing?

danlangford commented:
ok so some of the files in `./tools/python/user` still work with python 2 just fine

    cd ./tools/python/user/
    python monitor
    python game_info 80000

and in that directory my `.bmrc` file looks like this

```
[dev]
url=http://dev.buttonweavers.com/api/responder
username=
password=
cookiefile=.bmcookies.dev

[www]
url=http://www.buttonweavers.com/api/responder
username=Bagels
password=<REDACTED>
cookiefile=.bmcookies.www

```

there are some dependencies that you may need to install via pip. I dont know if that's your current issue I have them listed out in a couple `requirements.txt` files. im trying to decide where the best place for those files would be. from the `replaytest` dir you would do 

    pip install -r ../api_client/requirements.txt

or

    python -m pip install -r ../api_client/requirements.txt

cgolubi1 commented:
Again, the replay sites run puppet, so if puppet installs the dependencies, the replay sites should have them.  Anyway, no worries, i understand that it's on my plate to reverse-engineer what went wrong, i'll do it as soon as i can.

cgolubi1 commented:
Here's what i think the problem is: before this change, bmapi's code said:

```python
  def _setup_cookies(self):
    # all requests should use the same cookie jar
    self.cookiejar = LWPCookieJar(self.cookiefile)
    if os.path.isfile(self.cookiefile):
      self.cookiejar.load(ignore_discard=True)
```

That defined `self.cookiejar` whether or not the cookiefile existed on disk.

Now bmapi's code says:

```python
  def _setup_cookies(self):
    self.session = requests.session()
    # all requests should use the same cookie jar
    if self.cookiefile is not None and os.path.isfile(self.cookiefile):
      self.cookiejar = LWPCookieJar(self.cookiefile)
      self.cookiejar.load(ignore_discard=True)
      self.session.cookies = self.cookiejar
```

This defines `self.cookiejar` *only* if `self.cookiefile` already exists.  But that doesn't make sense --- cookiefile is a cache file.  If i define it and it doesn't exist yet, i want bmapi to create it for me.  And i definitely don't want bmapi to blow up with a mysterious stacktrace if i didn't touch a cache file before invoking it.

Go ahead and fix that, and i'll keep looking for more things.

cgolubi1 commented:
Further, *i* work with a cookiefile (i just don't touch it beforehand, because i'm standing up a fresh site every time), but i think your code is going to blow up the same way if someone uses a config file with no cookiefile too, right?  That's not okay --- if cookiefile is optional, we have to let people succeed in logging in without it.  And if it's not optional, we have to give people a clean error on init if they haven't defined it.

cgolubi1 commented on tools/python/random_ai/random_ai.py:
```diff
@@ -1706,7 +1707,7 @@ def record_create_game(self, button1, button2, max_wins=3, use_prev_game=False):
     else:
       custom_recipe_array = []
-    retval = self.player_client.create_game(button1, obutton=button2, opponent=self.opponent_client.username, max_wins=max_wins, use_prev_game=use_prev_game, custom_recipe_array=custom_recipe_array)
+    retval = tools.python.user.create_game(button1, obutton=button2, opponent=self.opponent_client.username, max_wins=max_wins, use_prev_game=use_prev_game, custom_recipe_array=custom_recipe_array)
```
This change seems incorrect --- it doesn't work:

```python
Starting game 1
Traceback (most recent call last):
  File "./test_log_games", line 71, in <module>
    c.log_test_game(i, button1, button2)
  File "../random_ai/random_ai.py", line 1802, in log_test_game
    self.record_create_game(button1, button2, use_prev_game=use_prev_game)
  File "../random_ai/random_ai.py", line 1710, in record_create_game
    retval = tools.python.user.create_game(button1, obutton=button2, opponent=self.opponent_client.username, max_wins=max_wins, use_prev_game=use_prev_game, custom_recipe_array=custom_recipe_array)
NameError: global name 'tools' is not defined
```

and it shouldn't be necessary --- `self.player_client` is just as defined as it ever was.

danlangford replied:
I think this was my refactor/rename tooling trying to help me when I moved a directory. a bit frustrating for sure. thanks for bringing it up.

cgolubi1 commented:
In addition to the two fixes described, i needed another change, because i modified things to run from the `.../replaytest` directory rather than the `...` directory, since that seemed to work more easily with your new code layout:

```python
$ git diff
diff --git a/tools/api-client/python/lib/random_ai.py b/tools/api-client/python/lib/random_ai.py
index f04baf67..9f1dc10b 100755
--- a/tools/api-client/python/lib/random_ai.py
+++ b/tools/api-client/python/lib/random_ai.py
@@ -1694,7 +1694,7 @@ class LoggingBMClient():
     self.log.append({
       'type': 'finish',
     })
-    f = open('output/game%s.pck' % self.int_id, 'w')
+    f = open('../output/game%s.pck' % self.int_id, 'w')
     pickle.dump(self.log, f)
     f.close()
 
```

That's gross --- obviously we should be instantiating random_ai with an output directory as a variable, and i'll do that now that it's drawn to my attention.  But the diff i have in this comment isn't any *worse* than what was there before, and it'll let this PR not break replaytest, so i think you should just do it (file an issue if it makes you sad, or i can), and we'll clean up later.

cgolubi1 commented:
So basically with those three changes:
1. make bmapi not blow up with stacktraces in either of the cases A. cookiefile is defined but doesn't exist yet (this case MUST be supported), and B. cookiefile is not defined (afaik this case can be supported, in which case it should be, but if it's not easy and you want to fail with a clean config error, that's okay with me)
2. don't make that stray one-line change in random_ai.py
3. do change random_ai.py to expect output to go into `../output`, basically Because I Say So, and sorry about that

i was able to update my replay site builder script to work with the new file locations, and it's running now.  Go ahead and fix those three things in the code, and i'll continue to let replay-testing look for subtler regressions.

cgolubi1 commented:
Oh, hmm, there's a problem with prepping games for replay too, give me 1s to describe and propose a fix.

danlangford commented:
I wasn't sure how the replay test stuff was working. I didn't realize it was run from each site, from the same server (I mean I guess that makes sense I just hadn't thought about it) that puppet is run on

ok ill work on

- [ ] cookie jar depending on cookie file
- [ ] cookie file in the config optional or required?
- [ ] self.player_client vs tools.python.user
- [ ] random_ai output directory

danlangford commented:
> Further, i work with a cookiefile
hey @cgolubi1 what do you do with cookiefile/cookiejar? I moved to using an in-memory cookie jar if a file isn't specified. so some API methods differ from CookieJar to File/LWPCookieJar.

cgolubi1 commented:
> hey @cgolubi1 what do you do with cookiefile/cookiejar? anything more than specifying / managing it?

No, less than that in fact --- i just specify it in my config file, but don't do anything to "manage" it --- i count on bmapi to log me in, i never interact with the cookiefile/cookiejar directly.  I'm pretty sure your change broke all users specifying a new cookiefile for the first time (you didn't see it in testing because you were testing with a cookiefile which already existed); there's nothing special about random_ai there.

The only thing i do that's potentially weird is that i'm playing both sides of the game from the same python code, so i have two BMClient instances hanging around, one logged in as `responder003` and one logged in as `responder004`.  That works just fine, and i don't see why anything in this refactor should break it (and it was working just fine with your revision once i patched in a fix for the cookiefile thing), i'm just mentioning it in case it comes up later.

cgolubi1 commented on tools/python/api_client/bmapi.py:
```diff
@@ -67,27 +43,38 @@ def __init__(self, response_dict):
-    self.cookieprocessor = HTTPCookieProcessor(self.cookiejar)
-    self.opener = build_opener(self.cookieprocessor)
-    install_opener(self.opener)
+    # ref session so requests can share session config like a shared cookiejar
+    self.session = requests.session()
+    if self.cookiefile is not None:
+      print('file cookies')
+      self.cookiejar = LWPCookieJar(self.cookiefile)
+      if os.path.isfile(self.cookiefile):
+        print('cookie cache')
+        # load up existing cachefile
+        self.cookiejar.load(ignore_discard=True)
+    else:
+      print('in mem cookies')
+      # use in-memory cookie jar if needed
+      self.cookiejar = CookieJar()
+    self.session.cookies = self.cookiejar
```
This seems reasonable, but i think you left some debugging print statements in.

danlangford replied:
hahaha I sure did. thanks for catching this.

cgolubi1 commented:
Okay, i think i finally solved the replay testing problem i was having (tl;dr the pickle files i use to gather data about games won't be backwards-compatible across this change because the location of bmapi changed, but that's not a problem as long as i have way to hack across it for testing *this* change, which i now do:

```python
#!/usr/bin/python
##### update_replay_games
# This should have the same functionality as prep_replay_games, but
# can be modified to adjust the API output to account for changes
# between the master version against which games were recorded, and
# the version under test.  One-off changes which will not be needed
# once the API update is committed should be made in this file and
# should not be committed.  Support for permanent API changes should
# be made in random_ai.py.

import sys
sys.path.append('..')

from random_ai import random_ai
from api_client import bmapi

sys.modules['bmapi'] = bmapi

########################################################################
# Custom classes
class PHPBMClientOutputWriter(random_ai.PHPBMClientOutputWriter):
  # If you need to make any changes to the syntax of PHP output which is
  # translated from the internal game state format for regression-testing
  # previously recorded games against this PR (e.g. if the new code
  # changes the API and you need to reflect those changes for replays
  # to succeed), make those overrides here for clarity during testing.
  #
  # As a first resort, trying overriding apply_known_changes_to_game_data()
  pass

########################################################################
# Initial setup
if len(sys.argv) < 2:
  raise ValueError("Usage: update_replay_games <dirname>")

DIRNAME = sys.argv[1]

d = PHPBMClientOutputWriter(DIRNAME)
for gamepath in d.find_games():
  sys.stderr.write("Updating and prepping game for replay from: %s\n" % gamepath)
  d.translate_game(gamepath)
```

(That doesn't get committed, it's a one-off change as noted in the file header comment.)  So i'll see how all of that goes.

I'm not really sure why github can't handle the updates to random_ai.py and is asking you to rebase, but FYI it is, and you can do that at your leisure; no rush, in case we make other unrelated random_ai changes for other PRs.

danlangford commented:
okie dokie the rebase is done. 

and I know I have just ditched the community and I won't be hurt if you guys drop this PR. but it may still benefit @irilyth and anybody else wanting to participate in the python space to be up on more current versions and standards

cgolubi1 commented:
Thanks very much for updating this, Dan.  I'll take a look as soon as i can.

cgolubi1 commented:
Okay, so at this point i'd say i've borrowed and adapted the following aspects of this CR, and gotten them merged under other CRs:
* Using a requirements file for python2 and python3 PIP requirements
* Installing standalone python2 and python3 versions that can be managed independently of the system python (i used conda, but the upshot is the same)
* Rearranging files a bit so that the replaytest scripts are sequestered from everything else, and adding some READMEs
* Running python2 and python3 unit tests, and updating bmutils and the CLI scripts so they work with python2 and python3 (i also added unit tests for bmutils and minimal "does it run at all in this python version?" unit tests for the standalone scripts)

I haven't done:
* A more in-depth directory rearrangement of the scripts and tests --- it'd be a fair amount of work to adapt the replay tests to, and i don't think the cost/benefit comes out.  If you're blocked from some particular project by where the files are in git, say more about how it's blocking you, and let's see what we can do.
* The python3-compatibility-ish update of the various replaytest scripts --- i can do that later when/if i update replay testing to python3

I feel reasonably happy with the current situation, and i think what i've done is sufficient to address #2197 (i'll raise that topic on that issue) and unblock #2668 (which is honestly pretty simple, and we should probably just do).

So i'm going to go ahead and close this out, and thank Dan *very* much for providing the incentive and making this code available to cannabalize!

danlangford commented:
@cgolubi1 I just merged your changes into my working directory. I think what you landed on is a great compromise and definitely unblocks a lot of things id like to contribute. thank you
